Clean up the pixel vertex reconstruction code

ericcano opened the pull request:
#### PR description:

This PR will implement remarks for https://github.com/cms-sw/cmssw/pull/31723.

#### PR validation:

The baseline and new code will be run through TTbar validations:

   -  [501: CPU - Quad+Pentaplets](https://patatrack.web.cern.ch/patatrack/validation/pulls/13a7a77294b8e32b503b6abf63fcf96c22939c56/RelValTTbar_14TeV-CMSSW_11_2_0_pre7-PU_112X_mcRun3_2021_realistic_v8-v1/development-11634.501-step3.py)
   -  [502: GPU - Quad+Pentaplets](https://patatrack.web.cern.ch/patatrack/validation/pulls/13a7a77294b8e32b503b6abf63fcf96c22939c56/RelValTTbar_14TeV-CMSSW_11_2_0_pre7-PU_112X_mcRun3_2021_realistic_v8-v1/development-11634.502-step3.py)

  - [505: CPU - Trips+Quads+Penta](https://patatrack.web.cern.ch/patatrack/validation/pulls/13a7a77294b8e32b503b6abf63fcf96c22939c56/RelValTTbar_14TeV-CMSSW_11_2_0_pre7-PU_112X_mcRun3_2021_realistic_v8-v1/development-11634.505-step3.py)
   - [506: GPU - Trips+Quads+Penta](https://patatrack.web.cern.ch/patatrack/validation/pulls/13a7a77294b8e32b503b6abf63fcf96c22939c56/RelValTTbar_14TeV-CMSSW_11_2_0_pre7-PU_112X_mcRun3_2021_realistic_v8-v1/development-11634.506-step3.py)

ericcano committed:
Fixed include guard naming.

https://github.com/cms-sw/cmssw/pull/31723#discussion_r545031685

ericcano committed:
Fixed member variable naming.

https://github.com/cms-sw/cmssw/pull/31723#discussion_r545033873
https://github.com/cms-sw/cmssw/pull/31723#discussion_r545034002
https://github.com/cms-sw/cmssw/pull/31723#discussion_r545034081

ericcano committed:
Separated produce() in 2 CPU/GPU functions.

Enclosed commented out code in ifdefs.

https://github.com/cms-sw/cmssw/pull/31723#discussion_r545034278
https://github.com/cms-sw/cmssw/pull/31723#discussion_r545034738

ericcano committed:
Simplified event data access.

Protected stdout printouts in ifdefs.

https://github.com/cms-sw/cmssw/pull/31723#discussion_r545035057
https://github.com/cms-sw/cmssw/pull/31723#discussion_r545035370
https://github.com/cms-sw/cmssw/pull/31723#discussion_r545037916
https://github.com/cms-sw/cmssw/pull/31723#discussion_r545038383

ericcano committed:
Fixed include guard naming.

https://github.com/cms-sw/cmssw/pull/31723#discussion_r545047024

ericcano committed:
Enclosed debug printouts in ifdefs.

https://github.com/cms-sw/cmssw/pull/31723#discussion_r545066158
https://github.com/cms-sw/cmssw/pull/31723#discussion_r545066219

ericcano committed:
Stored constants in variables.

https://github.com/cms-sw/cmssw/pull/31723#discussion_r545067045

ericcano committed:
Replaced relative paths.

https://github.com/cms-sw/cmssw/pull/31723#discussion_r549374469

fwyzard commented on RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc:
```diff
@@ -127,6 +129,7 @@ void PixelVertexProducerFromSoA::produce(edm::StreamID streamID, edm::Event &iEv
         edm::LogWarning("PixelVertexProducer") << "oops track " << it << " does not exists on CPU " << k;
         continue;
       }
+      auto trackCollection = iEvent.getHandle(tokenTracks_);
```
This is cheap, but not free.
Could you move it back outside of the loop ?

ericcano replied:
Sure, fixed in latest commits.

fwyzard commented on RecoPixelVertexing/PixelVertexFinding/plugins/gpuVertexFinderImpl.h:
```diff
@@ -112,37 +118,42 @@ namespace gpuVertexFinder {
 #ifdef __CUDACC__
+    // Running too many thread lead to problems when printf is enabled.
+    constexpr int maxThreadsForPrint = 1024 - 256;
+    constexpr int splitBlocks = 1024;
+    constexpr int splitThreadsPerBlock = 128;
```
could you use the names
```suggestion
    constexpr int numBlocks = 1024;
    constexpr int threadsPerBlock = 128;
```
?

ericcano replied:
Sure, fixed in latest commits.

ericcano committed:
Moved event data handle creation out of a loop.

https://github.com/cms-patatrack/cmssw/pull/609#discussion_r601259643

ericcano committed:
Renamed threading variables.

https://github.com/cms-patatrack/cmssw/pull/609#discussion_r601261542

fwyzard commented on RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc:
```diff
@@ -89,9 +87,12 @@ void PixelVertexProducerFromSoA::produce(edm::StreamID streamID, edm::Event &iEv
+#endif // PIXVERTEX_DEBUG_PRODUCE 
 
   std::set<uint16_t> uind;  // fort verifing index consistency
+  auto trackCollection = iEvent.getHandle(tokenTracks_);
```
sorry, rereading the changes, isn't this the same as `tracks` at line 67 ?

ericcano replied:
Indeed. 

We absolutely need a handle for making a TrackRef, as edm::Ref does not take references to product as a constructor parameter. It does does take pointers to product (and this compiles):
```c++
      //(old) auto tk = reco::TrackRef(trackCollectionHandle, k);
      auto tk = reco::TrackRef(&tracks, k);
```

but the constructor comments advise against using this to persist products which are in events (`DataFormats/Common/interface/Ref.h`):
```c++
    /// Constructors for ref to object that is not in an event.
    //  An exception will be thrown if an attempt is made to persistify
    //  any object containing this Ref.  Also, in the future work will
    //  be done to throw an exception if an attempt is made to put any object
    //  containing this Ref into an event(or run or lumi).
    Ref(product_type const* product, key_type itemKey, bool setNow = true);
```

So we can replace `track` with a handle and dereference it where needed (only one place, actually).

Fixed in latest commit, with an extra (unrelated) handle creation simplification.

fwyzard replied:
> It does take pointers to product (and this compiles)

Sure, but as you have seen in the comment that will give a transient-only `edm::Ref`, which does not know the original collection - so it's not what we want.

ericcano committed:
Removed duplicate product/handle.

Also simplified handle creation.

https://github.com/cms-patatrack/cmssw/pull/609#discussion_r601279027

ericcano committed:
Fixed code format.

fwyzard commented on RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc:
```diff
@@ -62,15 +64,9 @@ void PixelVertexProducerFromSoA::fillDescriptions(edm::ConfigurationDescriptions
-
-  edm::Handle<reco::BeamSpot> bsHandle;
-  iEvent.getByToken(tokenBeamSpot_, bsHandle);
+  auto tracksHandle = iEvent.getHandle(tokenTracks_);
```
```suggestion
  auto tracksHandle = iEvent.getHandle(tokenTracks_);
  const auto tracksSize = tracksHandle->size();
```

fwyzard commented on RecoPixelVertexing/PixelVertexFinding/plugins/PixelVertexProducerFromSoA.cc:
```diff
@@ -123,11 +124,11 @@ void PixelVertexProducerFromSoA::produce(edm::StreamID streamID, edm::Event &iEv
       assert(it < int(indToEdm.size()));
       auto k = indToEdm[it];
-      if (k > tracks.size()) {
+      if (k > tracksHandle->size()) {
```
```suggestion
      if (k > tracksSize) {
```

ericcano replied:
Indeed.

fwyzard reviewed (commented):
Last minor suggestion.

ericcano committed:
Avoided multiple access to handle.

https://github.com/cms-patatrack/cmssw/pull/609#discussion_r601377597

fwyzard commented:
:construction: Validation running at fu-c2a02-37-02:/data/user/fwyzard/patatrack/validation/run_609.720XqNxPB7 ...
